### DISCUSSION TITLE
start trial license without user interaction

### DISCUSSION THREAD

**Discussion by nschloe** [keygen-sh/community#17] [Q&A] [answered]
I would like to get users started with zero interaction, i.e., upon first start, initialize a trial license that expires after _n_ weeks.

How to best go about this?

- Create a dummy user and assign a new license to it every time someone starts for software for the first time?
- Create a new user every time and assign a new trial license to it. (Not sure how to do that though. No interaction should be required.)

Perhaps there are better alternatives.

**Comment by ezekg** [comment]
In this scenario, what would stop a user from possibly requesting more than 1 trial by simply generating a new dummy user? Typically, you would create a license for each trial user (again, typically server-side), using a trial policy that expires after a certain amount of time. Same flow for a full license. The trial user would input their key like any other licensed user. Then to limit duplicate trials, you would also want to activate their device and the fingerprint's unique across the trial policy's licenses (i.e. assert that the same device hasn't been activated for another trial license in the past, indicating a duplicate trial.)

I don't have a sure-fire solution for a 100% no-interaction licensing flow, but if you or anybody else has ideas, I'm all ears. I'll think on it more and see if I can come up with anything. Are you wanting to do everything client-side, or is a server-side flow OK?

**Reply by nschloe** [reply]
> Then to limit duplicate trials, you would also want to activate their device and the fingerprint's unique across the trial policy's licenses (i.e. assert that the same device hasn't been activated for another trial license in the past, indicating a duplicate trial.)

Yeah, I'd probably have to be tied to the machine. I don't see where I'd need a username or e-mail address at this point. The user could start the software for the first time, in the background it generates a machine-fingerprint (is that the way it works?), asks keygen to create a new trial license for a dummy user, and use that for _n_ weeks. After that, the user will have to provide her own license key, acquired in the meantime.

Do you see potential problems with this flow?

> Are you wanting to do everything client-side, or is a server-side flow OK?

I don't care either way. Offline is of course more robust, but I suppose I could retrofit a timeout like it's done for all other online licenses, too.

**Comment by ezekg** [comment, marked as answer]
You'd likely need to perform this custom flow server-side. You'd want to set up a server-side endpoint that accepts a fingerprint parameter, and either generates a license for it and activates the fingerprint, or retrieves an existing license, depending on if it's already activated. Licenses need to be explicitly created using the API, Keygen doesn't automatically create them for you.
